File: src/main/java/com/chaco/algorithms/tree/isSymmetric.java
package com.chaco.algorithms.tree;

import com.chaco.algorithms.TreeNode;

public class isSymmetric {
    public static void main(String[] args) {
        System.out.println("isSymmetric.main");
    }

    public boolean isSymmetric(TreeNode root) {
        return check(root, root);
    }

    public boolean check(TreeNode left, TreeNode right) {
        if (null == left && null == right) {
            return true;
        }
        if (null == left || null == right) {
            return false;
        }
        return left.val == right.val && check(left.left, right.right) && check(left.right, right.left);
    }
}

File: src/main/java/com/chaco/algorithms/tree/isSymmetric.go
package tree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	return check(root, root)
}

func check(left *TreeNode, right *TreeNode) bool {
	if nil == left && nil == right {
		return true
	}
	if nil == left || nil == right {
		return false
	}
	return left.Val == right.Val && check(left.Left, right.Right) && check(left.Right, right.Left)
}
